tactics: Display goal labels

mtzguido opened the pull request:
Support for a new feature incoming in Meta-F*: goal labels. They're just a way for users to keep track of the goals by adding custom messages, and get displayed when a goal fails to be solved by SMT as well. I just copied the formatting from the witness, but please feel free to choose a better one!

(I haven't yet looked at those other issues again.. sorry)

cpitclaudel commented on fstar-mode.el:
```diff
@@ -4854,7 +4859,8 @@ cell."
     (fstar--insert-ln-with-face 'fstar-goal-type-face (fstar-highlight-string .goal.type))
-    (fstar--insert-ln-with-face 'fstar-goal-witness-face (fstar-highlight-string .goal.witness))))
+    (fstar--insert-ln-with-face 'fstar-goal-witness-face (fstar-highlight-string .goal.witness))
+    (fstar--insert-ln-with-face 'fstar-goal-label-face (concat "Label:" (fstar-highlight-string .goal.label)))))
```
Suggestion: put the goal name before the goal, not after it: that is, move the goal name insertion to fstar-tactics--insert-goals.  For example something like Goal 1/3 (name here)

mtzguido commented:
Made these changes and updated the PR. I also changed F*'s CLI output to more closely match fstar-mode (which was undoubtedly doing a better job). Here's the test file:
```
module Labels

open FStar.Tactics

let _ = assert (True /\ (True /\ True)) by (explode (); iseq [(fun () -> tlabel "Left"); (fun () -> tlabel "Right")]; dump "dump msg")
```
The CLI output
```
proof-state: State dump @ depth 0 (dump msg):
Location: Labels.fst(5,118-5,133)
Goal 1/3 (Left):
 |- _ : Prims.squash Prims.l_True

Goal 2/3 (Right):
 |- _ : Prims.squash Prims.l_True

Goal 3/3:
 |- _ : Prims.squash Prims.l_True

Verified module: Labels (110 milliseconds)
All verification conditions discharged successfully
```
and emacs
![2018-08-29 17 33 06_35](https://user-images.githubusercontent.com/4195583/44822758-af18cf80-abb1-11e8-89ab-f493bd766d83.png)

Feel free to squash both commits into one, or I can do that too. Can you point me to the docs so I can update them?

cpitclaudel commented on fstar-mode.el:
```diff
@@ -4863,10 +4863,15 @@ cell."
-        kind goal-id ngoals (propertize "\n" 'line-spacing 0.2))
+        kind goal-id ngoals
+        (let ((lab (let-alist goal .goal.label)))
+             (if (equal lab "")
```
This looks wrong: with old version of F* `.goal.label` will be nil, not `""`

cpitclaudel commented on fstar-mode.el:
```diff
@@ -4863,10 +4863,15 @@ cell."
+        (let ((lab (let-alist goal .goal.label)))
+             (if (equal lab "")
+                 ""
+                 (concat " (" (concat lab ")"))))
```
You can give more than 2 arguments to `concat`: `concat " (" lab ")"`

cpitclaudel replied:
I'd write the whole thing like this:

```
(let ((label (let-alist goal .goal.label)))
  (if (member label '(nil "")) "" (format " (%s)" label)))
```

cpitclaudel replied:
Hmm, I don't see anything pushed to master :)

mtzguido replied:
These are the comments from a few months ago that I just submitted :). I was referring to commit fae299bfd8efaf50d03632351ceacd40d07d83f3 in `master`.

cpitclaudel replied:
Oh, woops. That commit looks fine to me :)

cpitclaudel commented:
Gentle ping. Did we merge this without closing the PR?

mtzguido commented:
Hey Clément. Indeed, I messed up and pushed to master instead of to my branch. I had left a comment in a review, but I now see I didn't hit "submit" so you couldn't see them, sorry :(.

cpitclaudel commented:
We're all good :)
